### PR TITLE
Geoid output

### PR DESCRIPTION
**Pull request by ian-r-rose**
Adds a visualization postprocessor for the geoid postprocessor.

### CODE REVIEW

File: include/aspect/postprocess/visualization/geoid.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
+/*
+  Copyright (C) 2016-217 by the authors of the ASPECT code.
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/geoid.h]: 2017

File: include/aspect/postprocess/visualization/geoid.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+           * will be ignored but must nevertheless exist in the returned
+           * vector. While implementations of this function must create this
+           * vector, ownership is taken over by the caller of this function
+           * and the caller will take care of destroying the vector pointed
+           * to.
+           */
+          virtual
+          std::pair<std::string, Vector<float> *>
+          execute () const;
+          /**
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/geoid.h]: please add an empty line before the comment

File: source/postprocess/geoid.cc [hunk 3]
```diff
@@ -446,8 +446,8 @@ namespace aspect
       std::vector<double> surface_dyna_topo_contribution_coesin; // a vector to store sin terms of surface dynamic topography contribution SH coefficients
       std::vector<double> CMB_dyna_topo_contribution_coecos; // a vector to store cos terms of CMB dynamic topography contribution SH coefficients
       std::vector<double> CMB_dyna_topo_contribution_coesin; // a vector to store sin terms of CMB dynamic topography contribution SH coefficients
-      std::vector<double> geoid_coecos;  // a vector to store cos terms of geoid anomaly SH coefficients
-      std::vector<double> geoid_coesin;  // a vector to store sin terms of geoid anomaly SH coefficients
+      geoid_coecos.resize(0);
+      geoid_coesin.resize(0);
```
Comment (gassmoeller) [inline source/postprocess/geoid.cc]: instead of `resize(0)` you can call `clear()`.

File: source/postprocess/visualization/geoid.cc [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                  unsigned int f = 0;
+                  for (; f<GeometryInfo<dim>::faces_per_cell; ++f)
+                    {
+                      if (cell->at_boundary(f) && this->get_geometry_model().depth (cell->face(f)->center()) < cell->face(f)->minimum_vertex_distance()/3)
+                        {
+                          surface_cell = true;
+                          break;
+                        }
+                    }
+
```
Comment (gassmoeller) [inline source/postprocess/visualization/geoid.cc]: please remove one empty line

File: source/postprocess/visualization/geoid.cc [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+                  //Get the location of the cell for the expansion
+                  const Point<dim> p = cell->face(f)->center();
+
+                  //Do the expansion at the point if it is a surface cell
```
Comment (gassmoeller) [inline source/postprocess/visualization/geoid.cc]: please add spaces between slashes and comment

File: source/postprocess/visualization/geoid.cc [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                    {
+                      if (cell->at_boundary(f) && this->get_geometry_model().depth (cell->face(f)->center()) < cell->face(f)->minimum_vertex_distance()/3)
+                        {
+                          surface_cell = true;
+                          break;
+                        }
+                    }
+
+
+                  //Get the location of the cell for the expansion
```
Comment (gassmoeller) [inline source/postprocess/visualization/geoid.cc]: see comment below

**Comment by ian-r-rose** [comment]
No, this test does not have negative densities.

**Comment by ian-r-rose** [comment]
I would like to apply a patch, but I don't see a diff from the tester here...

**Comment by tjhei** [comment]
Because the test crashes. Check the log.

**Comment by ian-r-rose** [comment]
Right you are. Should be fixed now.

File: source/postprocess/visualization/geoid.cc [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+// explicit instantiations
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(Geoid,
+                                                  "geoid",
+                                                  "Visualization for the geoid solution. The geoid is given"
```
Comment (tjhei) [inline source/postprocess/visualization/geoid.cc]: add space at end of line

File: source/postprocess/visualization/geoid.cc [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                                                  "by the equivalent water column height due to a gravity perturbation."
```
Comment (tjhei) [inline source/postprocess/visualization/geoid.cc]: and here

**Comment by ian-r-rose** [comment]
Spaces added.

**Comment by Shangxin-Liu** [comment]
Hi Ian, @ian-r-rose 

I have a question about your change of the geoid code to directly call dynamic topography post processor, which may lead to some further change of the comments. In the geoid.cc file, line 250 and 262, 

surface_stored_values.push_back (std::make_pair(fe_face_values.quadrature_point(q), std::make_pair(fe_face_values.JxW(q), topo_values[q])));

and

CMB_stored_values.push_back (std::make_pair(fe_face_values.quadrature_point(q), std::make_pair(fe_face_values.JxW(q), topo_values[q])));

Previously my original version stores the same cell average dynamic topography to each quadrature weight on the cell's boundary face in order to get higher transfer accuracy to spherical harmonic domain, as my above long comments explain. Now it seems that when calling your CBF dynamic topography directly, each quadrature point at one cell's boundary face may have different values of dynamic topography, yes? I need to update my comments if it is.

**Comment by ian-r-rose** [comment]
Yes, @Shangxin-Liu, different quadrature points on a cell face will, in general, have different values for dynamic topography.

**Comment by gassmoeller** [comment]
Feel free to discuss this further in a separate issue, but since this pull request is ready, and the discussion is unrelated to the code in this PR, I will go ahead and merge the new feature. Thanks Ian!

**Pull request merged**